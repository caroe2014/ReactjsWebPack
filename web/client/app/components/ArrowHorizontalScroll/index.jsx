// Copyright Â© 2018-2019 Agilysys NV, LLC.  All Rights Reserved.  Confidential Information of Agilysys NV, LLC.

import React, { Component, Fragment } from 'react';
import styled from 'styled-components';
import i18n from 'web/client/i18n';

const TabContainer = styled.div`
  display: flex;
  overflow: hidden;
  align-items: center;
  position: relative;
`;

const ScrollContent = styled.div`
  flex-direction: row;
  overflow-x: hidden;
  overflow-y: hidden;
  white-space: nowrap;
  display: flex;
  align-items: center;
  ${props => props.theme.mediaBreakpoints.tablet`
    overflow-x: auto;
  `};
  ${props => props.theme.mediaBreakpoints.mobile`
    overflow-x: auto;
  `};
 
`;

const RightArrow = styled.div`
  position: absolute;
  right: 0;
  z-index: 100;
  cursor: pointer;
  height: 100%;
  ${props => {
    if (props['visible']) {
      return `
      visibility: visible;
        opacity: 1;
      `;
    };
    return `
        visibility: hidden;
        opacity: 0;
    `;
  }};
   transition: opacity 200ms ease-in-out;
  @media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    top: 0;
  }
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

const LeftArrow = styled.div`
  position: absolute;
  z-index: 100;
  cursor: pointer;
  height: 100%;
  ${props => {
    if (props['visible']) {
      return `
      visibility: visible;
        opacity: 1;
      `;
    }
    return `
        visibility: hidden;
        opacity: 0;
    `;
  }};
  transition: opacity 200ms ease-in-out;
  @media screen and (-ms-high-contrast: active), (-ms-high-contrast: none) {
    top: 0;
  }
  &:focus {
    outline: none;
    box-shadow: none;
  }
`;

class ArrowHorizontalScroll extends Component {
  constructor (props) {
    super(props);
    this.state = {
      leftArrowVisible: false,
      rightArrowVisible: false
    };
    this.scrollInterval = this.scrollInterval.bind(this);
    this.onScroll = this.onScroll.bind(this);
    this.onContentScroll = this.onContentScroll.bind(this);
    this.handleKeyDownLeftArrow = this.handleKeyDownLeftArrow.bind(this);
    this.handleKeyDownRightArrow = this.handleKeyDownRightArrow.bind(this);
    this.scrollToDirection = this.scrollToDirection.bind(this);
  }

  componentDidMount () {
    if (this.modifierTab) {
      if (this.modifierTab.scrollWidth > this.modifierTab.offsetWidth) {
        this.setState({ rightArrowVisible: true });
      }
    }
  }

  scrollInterval (multipleOffset, scrollTo, direction) {
    this.currentPos += multipleOffset;
    this.modifierTab.scrollLeft = this.currentPos;
    if ((direction > 0 && parseInt(this.currentPos) >= scrollTo) ||
      (direction < 0 && parseInt(this.currentPos) <= scrollTo)) {
      clearInterval(this.intervalId);
      this.onContentScroll();
    }
  }

  onScroll = (direction) => (e) => {
    if (this.intervalId) {
      clearInterval(this.intervalId);
    }
    let offetWidth = this.modifierTab.offsetWidth / 2 * direction;
    this.currentPos = this.modifierTab.scrollLeft;
    let scrollTo = this.currentPos + offetWidth;
    let multipleOffset = Math.abs(offetWidth) > 100 ? offetWidth / 15 : offetWidth / 7;
    this.intervalId = setInterval(this.scrollInterval, 10, multipleOffset, scrollTo, direction);
  }

  scrollToDirection (direction) {
    if (this.intervalId) {
      clearInterval(this.intervalId);
    }
    let offetWidth = this.modifierTab.offsetWidth / 2 * direction;
    this.currentPos = this.modifierTab.scrollLeft;
    let scrollTo = this.currentPos + offetWidth;
    let multipleOffset = Math.abs(offetWidth) > 100 ? offetWidth / 15 : offetWidth / 7;
    this.intervalId = setInterval(this.scrollInterval, 10, multipleOffset, scrollTo, direction);
  }

  onContentScroll () {
    if (this.modifierTab.scrollLeft === 0) {
      this.setState({ leftArrowVisible: false, rightArrowVisible: true });
    } else if (Math.ceil(this.modifierTab.scrollLeft + this.modifierTab.offsetWidth) >= this.modifierTab.scrollWidth) {
      this.setState({ rightArrowVisible: false, leftArrowVisible: true });
    } else {
      this.setState({ leftArrowVisible: true, rightArrowVisible: true });
    }
  }

  handleKeyDownRightArrow (event) {
    // TODO: make this usable by all clickable elements,
    //  pass in caller's onclickhandler (13 is keycode for enter key, 32 is spacebar)
    if (event.which === 13 || event.which === 32) {
      this.scrollToDirection(1);
    }
  }

  handleKeyDownLeftArrow (event) {
    // TODO: make this usable by all clickable elements,
    //  pass in caller's onclickhandler (13 is keycode for enter key, 32 is spacebar)
    if (event.which === 13 || event.which === 32) {
      this.scrollToDirection(-1);
    }
  }

  render () {
    const { classContext, tabIndex } = this.props;
    return (
      <TabContainer className={classContext}>
        <Fragment>
          <LeftArrow
            className='scroll-arrow-left'
            visible={this.state.leftArrowVisible}
            onClick={this.onScroll(-1)}
            onKeyDown={this.handleKeyDownLeftArrow}
            tabIndex={this.state.leftArrowVisible && tabIndex}
            aria-label={i18n.t('SCROLL_LEFT')}
            role='button'
          >
            {this.props.leftArrow}
          </LeftArrow>
          <ScrollContent
            id='scroll-content'
            className='scroll-content'
            innerRef={(e) => { this.modifierTab = e; }}
            style={this.props.contentStyle}
            onScroll={this.onContentScroll}
          >
            {this.props.children}
          </ScrollContent>
          <RightArrow
            className='scroll-arrow-right'
            visible={this.state.rightArrowVisible}
            onClick={this.onScroll(1)}
            onKeyDown={this.handleKeyDownRightArrow}
            tabIndex={this.state.leftArrowVisible && tabIndex}
            aria-label={i18n.t('SCROLL_RIGHT')}
            role='button'
          >
            {this.props.rightArrow}
          </RightArrow>
        </Fragment>
      </TabContainer>
    );
  }
}

export default ArrowHorizontalScroll;
