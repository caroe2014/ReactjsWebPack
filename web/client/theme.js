// Copyright Â© 2018-2019 Agilysys NV, LLC.  All Rights Reserved.  Confidential Information of Agilysys NV, LLC.

import { css, injectGlobal } from 'styled-components';

const theme = {
  colors: {
    black: '#000000',
    navigationMenu: '#48a942',
    navigationMenuHover: 'cornflowerblue',
    primary: '#347CFF',
    primaryButton: '#347CFF',
    loginField: '#AAA7AC',
    placeHolder: '#909090',
    loginBoxColor: '#434149',
    open: '#48a942',
    close: 'red',
    footer: '#48a942',
    textGrey: '#505050',
    paymentSuccessTitle: '#505050',
    inputBottomBorder: 'lightgray',
    backgroundGrey: '#E6E6E6',
    text: 'black',
    primaryText: '#383838',
    secondaryText: '#a6a6a6',
    hint: '#a6a6a6',
    light: 'white',
    error: '#cc0000',
    tile: '#505050',
    addToCart: '#347CFF',
    itemTile: '#3C474C0',
    darkGrey: '#464646',
    itemBorder: 'lightgrey',
    addOnTab: '#347CFF',
    validationError: '#cc0000',
    loginBackground: '#303030',
    loginFormBackground: '#484848',
    disableBackground: '#E4E4E4',
    addonBg: '#EEEEEE',
    glutenFree: '#8d7356',
    spicy: '#ed1c24',
    vegan: '#0d4a22',
    vegetarian: '#0d4a22',
    new: '#ff6000',
    healthy: '#359919',
    contextBackground: '#f0f0f3',
    loader: '#ababab',
    disableButton: '#D1D1D1'
  },
  fontFamily: 'Roboto',
  receiptFontFamily: 'Roboto',
  fontSize: {
    xl: '2.4rem',
    lg: '1.8rem',
    md: '1.3rem',
    nm: '1rem',
    sm: '0.75rem'
  },
  fontSizeFromProps: (props) => {
    if (props.xl) return props.theme.fontSize.xl;
    if (props.lg) return props.theme.fontSize.lg;
    if (props.md) return props.theme.fontSize.md;
    if (props.nm) return props.theme.fontSize.nm;
    if (props.sm) return props.theme.fontSize.sm;
    return 'inherit';
  },
  spacing: {
    xl: '8em',
    lg: '4em',
    md: '2em',
    nm: '1em',
    sm: '0.5em',
    none: '0em'
  },
  mediaBreakpoints: {
    mobile: (...args) => css`@media (max-width: 32em) {${css(...args)}}`,
    tablet: (...args) => css`@media (max-width: 72em) {${css(...args)}}`,
    desktop: (...args) => css`@media (min-width: 72em) {${css(...args)}}`
  },
  shadow: '#cfcfcf 0.1px 0.1px 3px 0px',
  nav: {
    height: '70px'
  }
};

injectGlobal`
  .agilysys-icon-new {
    color: ${theme.colors.new}
  }
  .agilysys-icon-healthy {
    color: ${theme.colors.healthy}
  }
  .agilysys-icon-vegetarian {
    color: ${theme.colors.vegetarian}
  }
  .agilysys-icon-vegan {
    color: ${theme.colors.vegan}
  }
  .agilysys-icon-spicy {
    color: ${theme.colors.spicy}
  }
  .agilysys-icon-gluten-free {
    color: ${theme.colors.glutenFree}
  }
  @media print {
    @-moz-document url-prefix() {
      .appContainer, .appContainer>div, .appContainer .parent, .appContainer .BottomContainer {
        display: block !important;
      }
    }
  }
`;

export default theme;
