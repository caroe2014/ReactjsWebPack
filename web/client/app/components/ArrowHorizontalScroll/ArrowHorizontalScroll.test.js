// Copyright Â© 2018-2019 Agilysys NV, LLC.  All Rights Reserved.  Confidential Information of Agilysys NV, LLC.

import React from 'react';
import {
  shallow,
  mount,
  configure
} from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { ThemeProvider } from 'styled-components';
import theme from 'web/client/theme';
import 'jest-styled-components';

import ArrowHorizontalScroll from '.';
/* global describe, it, expect */
configure({
  adapter: new Adapter()
});

describe('ArrowHorizontalScroll', () => {

  it('does a snapshot test', () => {
    const wrapper = mount(<ThemeProvider theme={theme}><ArrowHorizontalScroll/></ThemeProvider>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render without throwing an error', () => {
    expect(shallow(<ArrowHorizontalScroll />).exists()).toBe(true);
  });

  it('renders right and left Arrows style rule', () => {
    let wrapper = shallow(<ArrowHorizontalScroll />);
    wrapper.setState({ rightArrowVisible: true, leftArrowVisible: true });
    wrapper.update();
    expect(wrapper.find('.scroll-arrow-left')).toHaveStyleRule('visibility', 'visible');
    expect(wrapper.find('.scroll-arrow-right')).toHaveStyleRule('visibility', 'visible');
  });

  it('click on scroll arrows', () => {
    let wrapper = mount(<ThemeProvider theme={theme}><ArrowHorizontalScroll /></ThemeProvider>);
    wrapper.find('.scroll-arrow-left').first().simulate('click');
    wrapper.find('.scroll-arrow-right').first().simulate('click');
    wrapper.find('.scroll-content').first().simulate('scroll');
    expect(wrapper.exists()).toBe(true);
  });

  it('should call  onScroll with mock', () => {
    let wrapper = shallow(<ArrowHorizontalScroll />);
    let mockFn = jest.fn();
    wrapper.instance().onScroll = mockFn;
    wrapper.instance().forceUpdate();
    wrapper.update();
    wrapper.find('.scroll-arrow-left').simulate('click');
    wrapper.find('.scroll-arrow-right').simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  it('should call  onContentScroll  with mock', () => {
    let wrapper = shallow(<ArrowHorizontalScroll />);
    let mockFn = jest.fn();
    wrapper.instance().onContentScroll = mockFn;
    wrapper.instance().forceUpdate();
    wrapper.update();
    wrapper.find('.scroll-content').simulate('scroll');
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('testing component did mount ref', () => {
    const myComponent = shallow(<ArrowHorizontalScroll/>, { disableLifecycleMethods: true });
    myComponent.instance().modifierTab = {
      scrollWidth: 100,
      offsetWidth: 2
    };
    myComponent.instance().componentDidMount();
  });

  it('testing component did mount ref', () => {
    const myComponent = shallow(<ArrowHorizontalScroll/>);
    myComponent.instance().modifierTab = {
      scrollWidth: 100,
      offsetWidth: 2
    };
    myComponent.instance().scrollInterval(10, 100, 1);
  });
});
