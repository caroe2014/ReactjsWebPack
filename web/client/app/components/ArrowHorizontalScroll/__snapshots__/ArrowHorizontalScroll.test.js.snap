// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowHorizontalScroll does a snapshot test 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c2 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  overflow-x: hidden;
  overflow-y: hidden;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  position: absolute;
  right: 0;
  z-index: 100;
  cursor: pointer;
  height: 100%;
  visibility: hidden;
  opacity: 0;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
}

.c3:focus {
  outline: none;
  box-shadow: none;
}

.c1 {
  position: absolute;
  z-index: 100;
  cursor: pointer;
  height: 100%;
  visibility: hidden;
  opacity: 0;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
}

.c1:focus {
  outline: none;
  box-shadow: none;
}

@media (max-width:72em) {
  .c2 {
    overflow-x: auto;
  }
}

@media (max-width:32em) {
  .c2 {
    overflow-x: auto;
  }
}

@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {
  .c3 {
    top: 0;
  }
}

@media screen and (-ms-high-contrast:active),(-ms-high-contrast:none) {
  .c1 {
    top: 0;
  }
}

<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "addOnTab": "#347CFF",
        "addToCart": "#347CFF",
        "addonBg": "#EEEEEE",
        "backgroundGrey": "#E6E6E6",
        "black": "#000000",
        "close": "red",
        "contextBackground": "#f0f0f3",
        "darkGrey": "#464646",
        "disableBackground": "#E4E4E4",
        "disableButton": "#D1D1D1",
        "error": "#cc0000",
        "footer": "#48a942",
        "glutenFree": "#8d7356",
        "healthy": "#359919",
        "hint": "#a6a6a6",
        "inputBottomBorder": "lightgray",
        "itemBorder": "lightgrey",
        "itemTile": "#3C474C0",
        "light": "white",
        "loader": "#ababab",
        "loginBackground": "#303030",
        "loginBoxColor": "#434149",
        "loginField": "#AAA7AC",
        "loginFormBackground": "#484848",
        "navigationMenu": "#48a942",
        "navigationMenuHover": "cornflowerblue",
        "new": "#ff6000",
        "open": "#48a942",
        "paymentSuccessTitle": "#505050",
        "placeHolder": "#909090",
        "primary": "#347CFF",
        "primaryButton": "#347CFF",
        "primaryText": "#383838",
        "secondaryText": "#a6a6a6",
        "spicy": "#ed1c24",
        "text": "black",
        "textGrey": "#505050",
        "tile": "#505050",
        "validationError": "#cc0000",
        "vegan": "#0d4a22",
        "vegetarian": "#0d4a22",
      },
      "fontFamily": "Roboto",
      "fontSize": Object {
        "lg": "1.8rem",
        "md": "1.3rem",
        "nm": "1rem",
        "sm": "0.75rem",
        "xl": "2.4rem",
      },
      "fontSizeFromProps": [Function],
      "mediaBreakpoints": Object {
        "desktop": [Function],
        "mobile": [Function],
        "tablet": [Function],
      },
      "nav": Object {
        "height": "70px",
      },
      "receiptFontFamily": "Roboto",
      "shadow": "#cfcfcf 0.1px 0.1px 3px 0px",
      "spacing": Object {
        "lg": "4em",
        "md": "2em",
        "nm": "1em",
        "none": "0em",
        "sm": "0.5em",
        "xl": "8em",
      },
    }
  }
>
  <ArrowHorizontalScroll>
    <styled.div>
      <div
        className="c0"
      >
        <styled.div
          aria-label="SCROLL_LEFT"
          className="scroll-arrow-left"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={false}
          visible={false}
        >
          <div
            aria-label="SCROLL_LEFT"
            className="scroll-arrow-left c1"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={false}
          />
        </styled.div>
        <styled.div
          className="scroll-content"
          id="scroll-content"
          innerRef={[Function]}
          onScroll={[Function]}
        >
          <div
            className="scroll-content c2"
            id="scroll-content"
            onScroll={[Function]}
          />
        </styled.div>
        <styled.div
          aria-label="SCROLL_RIGHT"
          className="scroll-arrow-right"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={false}
          visible={false}
        >
          <div
            aria-label="SCROLL_RIGHT"
            className="scroll-arrow-right c3"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={false}
          />
        </styled.div>
      </div>
    </styled.div>
  </ArrowHorizontalScroll>
</ThemeProvider>
`;
